Security considerations on untrusted input (#748)
openid/OpenID4VP: 1.1/openid-4-verifiable-presentations-1_1.md

MUST perform its own security checks on the returned Credentials and
Presentations.

## Parsing of untrusted inputs

> Wallets MUST treat all incoming requests as untrusted input. To mitigate injection and resource exhaustion attacks, Wallets MUST implement input validation on the Authorization Request and its enclosed DCQL query.

Reviewer: Is there a schema definition available for the authorization request and DCQL query?

Author: I wouldn't say strict schema, but we are defining the structures in the specification of both

Reviewer 2: Discussed this Q on today's APAC WG call. Leaving it for the larger group to discuss tomorrow:

1) Do we think JSON schemas can be useful here?

If there are no objections:
2) Do we think it's a 1.1 or 1.2 feature?

FYI @Sakurann @brentzundel

Reviewer 3: Thu wg discussion:
- no support to define JSON Schema: versioning, it is not standardized, security concerns for remote code execution, etc. 

text below might be tweaked better but reads good to strike the balance  
> * Enforce input validation: Verify that the Authorization Request and DCQL query contain no malformed properties. Unknown parameters MUST be ignored.

Reviewer 4: > Is there a schema definition available for the authorization request and DCQL query?

There's an (not officially working group endorsed, and not complete as it specialises for mdoc/sd-jwt vc) JSON schema for DCQL query available here: https://gitlab.com/openid/conformance-suite/-/tree/master/src/main/resources/json-schemas/oid4vp?ref_type=heads

Note that it's overly strict (rejects unknown properties, whereas spec requires them to be ignored) as we flag unknown properties in the conformance tests, as it tends to surface typos & misunderstandings of the spec.

> Wallets MUST treat all incoming requests as untrusted input. To mitigate injection and resource exhaustion attacks, Wallets MUST implement input validation on the Authorization Request and its enclosed DCQL query.

Reviewer 5: Should we mention the response for direct_post requests?

Wallets SHOULD implement the following steps:

> * Enforce Schema Validation: Verify that the DCQL query conforms to the expected JSON schema and contains no unknown or malformed properties.

Reviewer 6: I feel like that, in general, we need to drop unknown properties rather than fail on them. Inability to update client implementations along with us introducing new properties that only newer or a subset of wallets need seems important for practical inter-op. 

e.g. Imagine Wallet A supports a new format 'mso-zkp' which has some specific metadata parameters while Wallet B does not. A Verifier that wishes to request from both needs Wallet B to ignore the mso-zkp metadata parameters. 

Ideally we'd have clear rules for how dropping/failing works by having clearly defined extension points and typing, but as it stands I don't think this text is a good idea.

Author: Something like this? 

```suggestion
* Enforce Schema Validation: Verify that the Authorization Request and DCQL query contain no malformed properties. Unknown parameters MUST be ignored.
```

> 

Reviewer 5: ```suggestion
```

   * Clarified that Multi-RP-sig section means Verifier Info instead of attestations
   * Updated origin examples to remove trailing slash
   * Clarified that request_uri_method is a case-sensitive string

>    * add seecurity considerations on untrusted input

Reviewer 7: ```suggestion
   * add security considerations on untrusted input
```
